test/e2e/test_pull_subscription.go

```diff
@@ -122,3 +138,105 @@ func SmokePullSubscriptionTestImpl(t *testing.T) {
 		t.Error(err)
 	}
 }
+
+type TargetOutput struct {
+	Success bool `json:"success"`
+}
+
+// PullSubscriptionWithTargetTestImpl todo
+func PullSubscriptionWithTargetTestImpl(t *testing.T, packages map[string]string) {
+	topicName, deleteTopic := makeTopicOrDie(t)
+	defer deleteTopic()
+
+	psName := topicName + "-sub"
+	targetName := topicName + "-target"
+
+	client := Setup(t, true)
+	defer TearDown(client)
+
+	config := map[string]string{
+		"namespace":    client.Namespace,
+		"topic":        topicName,
+		"subscription": psName,
+		"targetName":   targetName,
+		"targetUID":    uuid.New().String(),
+	}
+	for k, v := range packages {
+		config[k] = v
+	}
+
+	installer := NewInstaller(client.Dynamic, config,
+		EndToEndConfigYaml([]string{"pull_subscription_target", "istio"})...)
+
+	// Create the resources for the test.
+	if err := installer.Do("create"); err != nil {
+		t.Errorf("failed to create, %s", err)
+		return
+	}
+
+	// Delete deferred.
+	defer func() {
+		if err := installer.Do("delete"); err != nil {
+			t.Errorf("failed to create, %s", err)
+		}
+		// Just chill for tick.
+		time.Sleep(10 * time.Second)
+	}()
+
+	gvr := schema.GroupVersionResource{
+		Group:    "pubsub.cloud.run",
+		Version:  "v1alpha1",
+		Resource: "pullsubscriptions",
+	}
+
+	jobGVR := schema.GroupVersionResource{
+		Group:    "batch",
+		Version:  "v1",
+		Resource: "jobs",
+	}
+
+	if err := client.WaitForResourceReady(client.Namespace, psName, gvr); err != nil {
+		t.Error(err)
+	}
+
+	topic := getTopic(t, topicName)
+
+	r := topic.Publish(context.TODO(), &pubsub.Message{
+		Attributes: map[string]string{
+			"target": "falldown",
+		},
+		Data: []byte(`{"foo":bar}`),
+	})
+	_, err := r.Get(context.TODO())
+	if err != nil {
+		t.Logf("%s", err)
+	}
+
+	msg, err := client.WaitUntilJobDone(client.Namespace, targetName)
+	if err != nil {
+		t.Error(err)
+	}
+	t.Logf("Last term message => %s", msg)
+
+	if msg != "" {
+		out := &TargetOutput{}
+		if err := json.Unmarshal([]byte(msg), out); err != nil {
+			t.Error(err)
+		}
+		if !out.Success {
+			// Log the output pull subscription pods.
+			if logs, err := client.LogsFor(client.Namespace, psName, gvr); err != nil {
+				t.Error(err)
+			} else {
+				t.Logf("pullsubscription: %+v", logs)
+			}
+			// Log the output of the target job pods.
+			if logs, err := client.LogsFor(client.Namespace, targetName, jobGVR); err != nil {
+				t.Error(err)
+			} else {
+				t.Logf("job: %s\n", logs)
+			}
+			t.Fail()
+		}
+	}
```

Comment:
oops failed to post this comment. Seems like if the msg == "", we should probably fail the test?